packages/@sanity/form-builder/src/inputs/ObjectInput/Field.tsx

```diff
@@ -1,101 +1,93 @@
+import {FormFieldPresence} from '@sanity/base/presence'
+import {Marker, Path, SchemaType} from '@sanity/types'
 import {FormBuilderInput} from '../../FormBuilderInput'
-import InvalidValue from '../InvalidValueInput'
-import styles from './styles/Field.css'
+import {InvalidValueInput} from '../InvalidValueInput'
+import PatchEvent from '../../PatchEvent'
 
-type FieldProps = {
-  field: any
-  value?: any
-  compareValue?: any
-  onChange: (...args: any[]) => any
-  onFocus: (...args: any[]) => any
-  onBlur: (...args: any[]) => any
-  focusPath?: any[]
-  filterField?: (...args: any[]) => any
+interface FieldType {
+  name: string
+  type: SchemaType
+}
+interface FieldProps {
+  field: FieldType
+  value: unknown
+  compareValue: unknown
+  onChange: (event: PatchEvent, field: FieldType) => void
+  onFocus: (path: Path) => void
+  onBlur: () => void
+  focusPath?: Path
+  filterField?: (type: SchemaType) => boolean
   readOnly?: boolean
-  markers?: any[]
-  level?: number
-  presence: any
+  markers?: Marker[]
+  level: number
+  presence?: FormFieldPresence[]
 }
 // This component renders a single type in an object type. It emits onChange events telling the owner about the name of the type
 // that changed. This gives the owner an opportunity to use the same event handler function for all of its fields
-export default class Field extends React.PureComponent<FieldProps> {
-  _input: any
-  static defaultProps = {
-    level: 0,
-    focusPath: [],
-  }
-  handleChange = (event) => {
-    const {field, onChange} = this.props
-    if (!field.type.readOnly) {
-      onChange(event, field)
-    }
-  }
-  focus() {
-    if (this._input && typeof this._input.focus === 'function') {
-      this._input.focus()
-    }
-  }
-  setInput = (input) => {
-    this._input = input
-  }
-  render() {
-    const {
-      value,
-      readOnly,
-      field,
-      level,
-      onFocus,
-      onBlur,
-      markers,
-      focusPath,
-      filterField,
-      compareValue,
-      presence,
-    } = this.props
-    if (typeof value !== 'undefined') {
-      const expectedType = field.type.name
-      const actualType = resolveTypeName(value)
-      // todo: we should consider removing this, and not allow aliasing native types
-      // + ensure custom object types always gets annotated with _type
-      const isCompatible = actualType === field.type.jsonType
-      if (expectedType !== actualType && !isCompatible) {
-        return (
-          <div className={styles.root}>
-            <Fieldset legend={field.type.title} level={level} presence={presence}>
-              <InvalidValue
-                value={value}
-                onChange={this.handleChange}
-                validTypes={[field.type.name]}
-                actualType={actualType}
-                ref={this.setInput}
-              />
-            </Fieldset>
-          </div>
-        )
+export const Field = forwardRef(function Field(props: FieldProps, forwardedRef: ForwardedRef<any>) {
```

Comment:
Probably not strictly in scope here, but would be nice to have a less general name for this component – e.g. `ObjectInputField`.

Reply:
Great suggestion. Will rename!

Reply:
Fixed in https://github.com/sanity-io/sanity/commit/b2f91754eac79fb9ef693cb362d4af0148568a25